crates/libs/implement/src/gen.rs

```diff
@@ -0,0 +1,579 @@
+        // for maintaining one of our invariants: Application code never has an owned instance
+        // of the outer (implementation) type. into_static() maintains this invariant by
+        // returning a wrapped StaticComObject value, which owns its contents but never gives
+        // application code a way to mutably access its contents. This prevents the refcount
+        // shearing problem.
+        //
+        // TODO: Make it impossible for app code to call this function, by placing it in a
+        // module and marking this as private to the module.
+        #[inline(always)]
+        #maybe_const fn into_outer(self) -> #impl_ident::#generics {
+            #impl_ident::#generics {
+                #initializers
+                count: ::windows_core::imp::WeakRefCount::new(),
+                this: self,
+            }
+        }
+    }
+}
+
+/// Generates the `T::into_static` function. This function is part of how we construct a
+/// `StaticComObject<T>` from a `T`.
+fn gen_into_static(inputs: &ImplementInputs) -> syn::ImplItem {
+    assert!(!inputs.is_generic);
+    parse_quote! {
+        /// This converts a partially-constructed COM object (in the sense that it contains
+        /// application state but does not yet have vtable and reference count constructed)
+        /// into a `StaticComObject`. This allows the COM object to be stored in static
+        /// (global) variables.
+        pub const fn into_static(self) -> ::windows_core::StaticComObject<Self> {
+            ::windows_core::StaticComObject::from_outer(self.into_outer())
+        }
+    }
+}
+
+/// Generates `From`-based conversions.
+///
+/// These conversions convert from the user's type `T` to `ComObject<T>` or to an interface
+/// implemented by `T`. These conversions are shorthand for calling `ComObject::new(value)`.
+///
+/// We can only generate conversions from `T` to the roots of each interface chain. We can't
+/// generate `From` conversions from `T` to an interface that is inherited by an interface chain,
+/// because this proc macro does not have access to any information about the inheritance chain
+/// of interfaces that are referenced.
+///
+/// For example:
+///
+/// ```rust,ignore
+/// #[implement(IFoo3)]
+/// struct MyType;
+/// ```
+///
+/// If `IFoo3` inherits from `IFoo2`, then this code will _not_ generate a conversion for `IFoo2`.
+/// However, user code can still do this:
+///
+/// ```rust,ignore
+/// let ifoo2 = IFoo3::from(MyType).into();
+/// ```
+///
+/// This works because the `IFoo3` type has an `Into` impl for `IFoo2`.
+fn gen_impl_from(inputs: &ImplementInputs) -> Vec<syn::Item> {
+    let mut items = Vec::new();
+
+    let original_ident = &inputs.original_type.ident;
+    let generics = &inputs.generics;
+    let constraints = &inputs.constraints;
+
+    items.push(parse_quote! {
+        impl #generics ::core::convert::From<#original_ident::#generics> for ::windows_core::IUnknown where #constraints {
```

Comment:
`QueryInterface` also supports `IAgileObject`, does that need to be added here as well?

Reply:
Probably, but I'll let Kenny speak to that.  It might be the case that we will always generate the vtable methods, but they'll only be wired up to QI when a user asks for it.

If we do change that, let's change it in a separate PR.

Reply:
While QI for `IAgileObject` is handled automatically by the `implement` macro, developers will never use an infallible conversion like `From` for `IAgileObject` since the whole point of `IAgileObject` is to ask the question to make a run time determination.

Reply:
I'm also wondering whether we really need these `From` implementation at all as its not really the most elegant way to construct a COM object any longer, but that is a discussion for another issue and PR.

Reply:
I'm fine with removing them.  I think `ComObject::new(foo).into()` is more clear and obvious than just `IFoo::from(foo)`, since it makes it clear that you're constructing a COM object and isn't much longer.

Will do in a separate PR.